program_intent_engagement/apps/api/v1/views.py

```diff
@@ -1 +1,61 @@
-# Create your views here.
+"""
+V1 API Views
+"""
+from rest_framework import status
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from django.db.models import F, Window
+from django.db.models.functions import RowNumber
+
+from program_intent_engagement.apps.core.models import ProgramIntent
+from program_intent_engagement.apps.api.serializers import ProgramIntentSerializer
+
+
+
+
+# TODO Do I add an @schema thing here?
+class MostRecentAndCertainIntentsView(APIView):
+    """
+    Retrieve a list of user intents
+    This endpoint returns a list of intents specified by the user.
+      Grouped by programs and sorted by most recent CERTAIN intent
+    Path: /api/[version]/intents/most-recent-and-certain
+    Returns:
+     * 200: OK
+     * 400: Query returned no results
+    """
+    authentication_classes = (JwtAuthentication,)
+    serializer_class = ProgramIntentSerializer
+
+    def get_queryset(self):
+        """
+        This view should return the best intents per program
+        for the currently authenticated user.
+
+        Best intent is defined as the most recent CERTAIN intent
+        """
+        user = self.request.user
+
+        result = (ProgramIntent.objects
+                      .filter(user=user)
+                      .order_by(F('effective_date'), F('certainty'))
+                      .annotate(
+                        row_number=Window(
+                          expression=RowNumber(),
+                          partition_by=[F('program_uuid')],
+                          ),
+                      )
+                      .filter(row_number = 1)
+                  )
+
+
+        if result is None:
```

Comment:
I don't think this belongs in a `get_queryset` method. `get_queryset` is intended to return the queryset, not to return a response.

Reply:
gooot it, so should i take everything out of the queryset then?

Reply:
I wouldn't expect this to work well as a generic view since it's not just directly listing a django model but it's worth a look and you'll learn something.

I would probably break out these methods:
get intents for user from DB - this probably doesn't even need a test, it's just a wrapper for base django. Maybe it doesn't even need a function.
filter a large pile intents into just the one you want for each program - just a loose function that can be tested by itself, this has interesting test cases
serializer for a pile of intents
view which figures out from the incoming user which local user it is (matching the ID in the JWT with the lms_user_id) and links together the others. It can have pretty simple tests.

Reply:
I think the [ListAPIView](https://www.django-rest-framework.org/api-guide/generic-views/#listapiview) would work, no? You can still override `get_queryset` to specify which model instances to return, and you can specify a serializer as well to pulls out the data that will be included in the response. The rest of the response handling is done by default by the ListAPIView. Or am I missing something?

Reply:
you're probably right, I'm just very biased to simple immediate code over inheritance. 
Marcin you're writing it, you can see how it feels to you. :)